### ISSUE TITLE
Encapsulate attributes for ExtensionContext

### ISSUE THREAD

**Issue by jlink** [junit-team/junit5#82]
Encapsulate attributes for Test|ContainerExtensionContext:
- do not expose map
- delegate attribute retrieval to parent

**Comment by jlink** [comment]
Implemented encapsulating methods in https://github.com/junit-team/junit-lambda/commit/cf91ce4cd20d11da24b139254d146571cdcbbc8e

**Comment by jlink** [comment]
Alternative suggestion: Implement 
```java
Store<T> ExtensionContext.getStore(Class<T> type, Scope scope)
```
with
```java
class Store<T> {
  T get() {}
  void set(T value) {}
```
and
```java
enum Scope {
  DEFAULT, //visible only in this extension but inherited from ancestor extension contexts
  LOCAL, //only visible in this extension and this extension context
  GLOBAL //visible in all extensions on this level and in ancestor extension contexts
}
```
Scope would probably use a key-value-based attributes mechanism underneath.

**Comment by jlink** [comment]
Team decision for the time being: 
Implement the following:
```java
Object get(Object key);

	private void put(Object key, Object value, Visibility visibility);

	default void put(Object key, Object value) {
		put(key, value, Visibility.DEFAULT);
	}

	private Object getOrComputeIfAbsent(Object key, Function<Object, Object> defaultProvider, Visibility visibility) {
		Object value = get(key);
		if (value == null) {
			value = defaultProvider.apply(key);
			put(key, value, visibility);
		}
		return value;
	}

	default Object getOrComputeIfAbsent(Object key, Function<Object, Object> defaultProvider) {
		return getOrComputeIfAbsent(key, defaultProvider, Visibility.DEFAULT);
	}
	enum Visibility {
		DEFAULT, //visible only in this extension but inherited from ancestor extension contexts
		//EXTENSION, //visible only in this extension but visible in the full tree of extension contexts
		//LOCAL, //only visible in this extension and for the current extension context
		//GLOBAL //visible in all extensions and in the full tree of extension contexts
	}
```

but hide Visibility from users for the time being until we've decided on the essential questions #112

**Comment by jlink** [comment]
I'm working on it.

**Comment by jlink** [comment]
I suggest three more changes in addition to the ones above:
- extract put/get/etc into an interface Store
- make the store available via ExtensionContext.getStore() 
- Alternative: Make it available as a separate parameter in all extension points to communicate: That's the way to persist state
- Encapsulate namespace in a Namespace object with utility methods like default(), local(), forClass() etc

Feedback?

**Comment by jlink** [comment]
What I'm planning to do is implement it in master but leave the old attributes in place. Thus you can all see how it looks but no existing functionality must be touched. 

Any objections?

**Comment by bechte** [comment]
For anything new features, for which we do not yet know in which direction we should go, I would strongly recommend not to work in master. That's exactly what I am doing with the discovery mechanism.

**Comment by jlink** [comment]
Ok. I'll only do in master on what we agreed yesterday.

**Comment by jlink** [comment]
Agreed changes are done in master. Suggestions from above must still be discussed.

**Comment by jlink** [comment]
Implemented suggestion in https://github.com/junit-team/junit5/tree/issue82-store for review

**Comment by jlink** [comment]
Merged suggestion after team decision in #128.

Missing: 
- [ ] Remove all usages of `ExtensionContext.put/get/removeAttribute`
- [ ] Remove implementation of `ExtensionContext.put/get/removeAttribute`

**Comment by jlink** [comment]
Resolved in https://github.com/junit-team/junit5/commit/e103a166d818926a53e0703b01abe6e99f349000
